rfcs/62-editor-data-model.md

```diff
@@ -0,0 +1,562 @@
+
+## User-facing explanation
+
+The Bevy Editor creates and edits Game data to build scenes and eventually organize the various Game assets into some cohesive game content.
+
+The _data model_ of the Editor refers to the way the Editor manipulates this Game data, their representation in memory while the Editor executable is running, and the API to do those manipulations. The Editor at its core defines a centralized data model, and all the ECS systems of the Editor use the same API to access that data model, making it the unique source of truth while the Editor process is running. This unicity prevents desynchronization between various subsystems, since they share a same view of the editing data. It also enables global support for common operations such as copy/paste and undo/redo without requiring each subsystem to implement their own variant.
+
+The Game uses Rust types defined in one of the `bevy_*` crates, any other third-party plugin library, or the Game itself (_custom types_). The data is represented in memory at runtime by _instances_ (allocated objects) of those Rust types. This representation is optimal in terms of read and write access speed, but it is _static_, that is the type layout is defined during compilation and cannot be changed afterward.
+
+In contrast, we want to enable the Editor to observe changes in the Game code without rebuilding or even restarting the Editor process, in order to enable fast iteration for the Authors. This feature, referred to as _hot-reloading_, requires the Editor to support type changes, which is not possible with static types like the Game uses. Therefore, to allow this flexibility, the Editor data model is based on _dynamic types_ defined exclusively during Editor execution. The data model defines an API to instantiate objects from dynamic types, manipulates the data of those instances, but also modify the types themselves while objects are instantiated. This latter case is made possible through a set of migration rules allowing to transform an object instance from a version of a type to another modified version of that type.
+
+The general workflow is as follows:
+
+- The Author starts the Editor executable and create a new project or loads an existing project.
+- The Editor launches the Game executable in _edit mode_, a special mode whereby the `EditorClientPlugin` is enabled and allows the Editor and the Game to communicate in real time. The `EditorClientPlugin` is a Bevy plugin (implements the `Plugin` trait) provided by an Editor crate and used during development only, which implements the Game side of the Editor/Game communication. It's compiled out of any Release build before shipping the Game.

```

Comment:
> It's compiled out of any Release build before shipping the Game.

This sentence is unclear, made me think the opposite when I first read it. “This plugin is not included in release builds of the Game.” sounds better imo.

Reply:
Fixed I think in new rewriting.